File: check_app_updates.py
import requests
import json
import os
import csv
import io
from google_play_scraper import app
from datetime import datetime

# ========== CONFIG ==========
DISCORD_WEBHOOK_URL = os.getenv('DISCORD_WEBHOOK_URL', 'https://discord.com/api/webhooks/xxx/yyy')
STATE_FILE = "last_versions.json"
APPS_SHEET_URL = os.getenv("APPS_SHEET_URL")  # <-- URL CSV Google Sheet (Publish to web)
APPS_CONFIG = os.getenv('APPS_CONFIG')
# ===========================


# ---------- Helpers for config ----------
def load_apps_config_from_sheet(url: str):
    """Đọc Google Sheet (CSV) thành dict {'ios': [], 'android': []}"""
    try:
        r = requests.get(url, timeout=15)
        r.raise_for_status()
        content = r.content.decode("utf-8", errors="ignore")
        reader = csv.DictReader(io.StringIO(content))

        cfg = {"ios": [], "android": []}
        for row in reader:
            platform = (row.get("platform") or "").strip().lower()
            app_id = (row.get("id") or "").strip()
            name_fallback = (row.get("name_fallback") or "").strip() or app_id

            # Skip blank rows / separators
            if not platform or not app_id:
                continue

            if platform in ("ios", "android"):
                cfg[platform].append({"id": app_id, "name_fallback": name_fallback})

        # Return None if sheet is effectively empty
        if not cfg["ios"] and not cfg["android"]:
            return None

        print(f"✅ Loaded apps config from Google Sheet (CSV): {len(cfg['ios'])} iOS, {len(cfg['android'])} Android")
        return cfg

    except Exception as e:
        print(f"❌ Error reading Google Sheet CSV: {e}")
        return None


def load_apps_config():
    """Priority: Google Sheet CSV -> APPS_CONFIG (JSON) -> hardcoded fallback."""
    # 1) Try Google Sheet CSV
    if APPS_SHEET_URL:
        cfg = load_apps_config_from_sheet(APPS_SHEET_URL)
        if cfg:
            return cfg
        print("⚠️ Google Sheet invalid/empty. Falling back to APPS_CONFIG.")

    # 2) Try APPS_CONFIG JSON env
    if APPS_CONFIG:
        try:
            cfg = json.loads(APPS_CONFIG)
            print(f"✅ Loaded apps config from APPS_CONFIG (JSON): {len(cfg.get('ios', []))} iOS, {len(cfg.get('android', []))} Android")
            return cfg
        except Exception as e:
            print(f"❌ Lỗi khi parse APPS_CONFIG: {e}")
            print(f"Config string: {APPS_CONFIG[:100]}...")

    # 3) Fallback
    print("⚠️ Using hardcoded fallback config.")
    return {
        "ios": [
            {"id": "1517783697", "name_fallback": "Genshin Impact"}
        ],
        "android": [
            {"id": "com.miHoYo.GenshinImpact", "name_fallback": "Genshin Impact"}
        ]
    }


# ---------- State ----------
def load_state():
    try:
        if os.path.exists(STATE_FILE):
            with open(STATE_FILE, "r") as f:
                return json.load(f)
        return {}
    except Exception as e:
        print(f"❌ Lỗi khi load state: {e}")
        return {}


def save_state(state):
    try:
        with open(STATE_FILE, "w") as f:
            json.dump(state, f, indent=2)
        print(f"✅ Đã lưu state vào {STATE_FILE}")
    except Exception as e:
        print(f"❌ Lỗi khi save state: {e}")

# ---------- Fetchers ----------
def get_ios_info(app_id):
    try:
        url = f"https://itunes.apple.com/lookup?id={app_id}"
        res = requests.get(url, timeout=15).json()
        if "results" in res and len(res["results"]) > 0:
            app_data = res["results"][0]
            return {
                "name": app_data["trackName"],
                "version": app_data["version"],
                "url": app_data["trackViewUrl"],
                "icon": app_data["artworkUrl100"],
                "releaseNotes": app_data.get("releaseNotes", "Không có ghi chú"),
                "releaseDate": format_date(app_data.get("currentVersionReleaseDate")),
                "developer": app_data.get("artistName", "Unknown"),
                "developerUrl": app_data.get("artistViewUrl", app_data["trackViewUrl"]),  # link dev riêng
                "genres": ", ".join(app_data.get("genres", [])) or app_data.get("primaryGenreName", "Unknown"),
                "screenshot": (app_data.get("screenshotUrls") or [None])[0],
            }
        return None
    except Exception as e:
        print(f"❌ Lỗi khi lấy thông tin iOS app {app_id}: {e}")
        return None

def get_android_info(pkg_name):
    try:
        result = app(pkg_name)
        return {
            "name": result["title"],
            "version": result["version"],
            "url": result["url"],
            "icon": result["icon"],
            "releaseNotes": result.get("recentChanges", "Không có ghi chú"),
            "releaseDate": format_date(result.get("updated")),
            "developer": result.get("developer", "Unknown"),
            "developerUrl": f"https://play.google.com/store/apps/dev?id={result.get('developerId')}" if result.get("developerId") else result["url"],
            "genres": result.get("genre", "Unknown"),
            "screenshot": (result.get("screenshots") or [None])[0],
        }
    except Exception as e:
        print(f"❌ Lỗi khi lấy thông tin Android app {pkg_name}: {e}")
        return None

# ---------- Discord ----------
def send_discord_embed(app_name, platform, old_version, info):
    try:
        embed = {
            "title": f"📢 {app_name} {platform} vừa cập nhật!",
            "url": info["url"],
            "description": (
                f"**Platform:** {platform}\n"
                f"**Old Version:** `{old_version or 'N/A'}`\n"
                f"**New Version:** `{info['version']}`\n"
                f"**Release Date:** {info.get('releaseDate', 'Không rõ')}\n"
                f"**Developer:** [{info.get('developer','Unknown')}]({info.get('developerUrl', 'url')})\n"
                f"**Genres:** {info.get('genres','Unknown')}\n"
                f"\n"
                f"**Release Notes:**\n{info['releaseNotes'][:1000]}"
            ),
            "color": 0x1abc9c,
            "thumbnail": {"url": info["icon"]},
            "footer": {"text": "App Update Notifier"}
        }

        if info.get("screenshot"):
            embed["image"] = {"url": info["screenshot"]}

        payload = {"embeds": [embed]}
        response = requests.post(DISCORD_WEBHOOK_URL, json=payload, timeout=15)
        response.raise_for_status()
        print(f"✅ Đã gửi thông báo Discord cho {app_name}")
    except Exception as e:
        print(f"❌ Lỗi khi gửi Discord: {e}")

# ---------- Utils ----------
def format_date(value):
    """Chuẩn hóa các kiểu date thành DD-MM-YYYY"""
    if not value:
        return "Không rõ"

    # Nếu là datetime hoặc date object
    if hasattr(value, "strftime"):
        return value.strftime("%d-%m-%Y")

    # Nếu là int (timestamp ms hoặc s)
    if isinstance(value, int):
        # Nếu lớn hơn năm 3000 thì chắc chắn ms
        ts = value / 1000 if value > 32503680000 else value
        return datetime.fromtimestamp(ts).strftime("%d-%m-%Y")

    # Nếu là string ISO (iOS thường có dạng 2024-08-12T10:15:23Z)
    if isinstance(value, str):
        try:
            return datetime.fromisoformat(value.replace("Z", "")).strftime("%d-%m-%Y")
        except Exception:
            # Nếu dạng "August 25, 2025"
            try:
                return datetime.strptime(value, "%B %d, %Y").strftime("%d-%m-%Y")
            except Exception:
                return value  # fallback giữ nguyên

    return str(value)

# ---------- Main ----------
def main():
    print("🔄 Bắt đầu kiểm tra cập nhật...")
    APPS = load_apps_config()
    print(f"📋 Loaded config: {len(APPS.get('ios', []))} iOS apps, {len(APPS.get('android', []))} Android apps")
    
    state = load_state()
    has_update = False

    print(f"📱 Đang kiểm tra iOS apps...")
    for app_data in APPS.get("ios", []):
        print(f"   🔍 Kiểm tra {app_data['name_fallback']}...")
        info = get_ios_info(app_data["id"])
        if info:
            old_version = state.get(app_data["id"])
            print(f"    - Cũ: {old_version or 'N/A'} | Mới: {info['version']} | Khác nhau: {info['version'] != old_version")
            if info["version"] != old_version:
                print(f"   🎉 PHÁT HIỆN CẬP NHẬT: {info['name']} | {old_version} → {info['version']}")
                send_discord_embed(info['name'], "iOS", old_version, info)
                state[app_data["id"]] = info["version"]
                has_update = True
            else:
                print(f"   ✅ {info['name']} (iOS) không có cập nhật")

    print(f"🤖 Đang kiểm tra Android apps...")
    for app_data in APPS.get("android", []):
        print(f"   🔍 Kiểm tra {app_data['name_fallback']}...")
        info = get_android_info(app_data["id"])
        if info:
            old_version = state.get(app_data["id"])
            print(f"    - Cũ: {old_version or 'N/A'} | Mới: {info['version']} | Khác nhau: {info['version'] != old_version}")
            if info["version"] != old_version:
                print(f"   🎉 PHÁT HIỆN CẬP NHẬT: {info['name']} | {old_version} → {info['version']}")
                send_discord_embed(info['name'], "Android", old_version, info)
                state[app_data["id"]] = info["version"]
                has_update = True
            else:
                print(f"   ✅ {info['name']} (Android) không có cập nhật")

    if has_update:
        save_state(state)
        print("💾 Đã lưu state mới")
    else:
        print("ℹ️ Không có cập nhật nào")

if __name__ == "__main__":
    main()

File: check_app_updates.js
// check_app_updates.js
import fs from "fs";
import axios from "axios";
import Papa from "papaparse";
import * as cheerio from "cheerio";
import { createThreadInTextChannel, sendMessageToThread, unarchiveThread } from "./helpers/discord_bot.js";
import { pickChannelId, ensureThreadBelongsToChannel, reuseThreadByNameInThisChannel } from "./helpers/discord_channel.js";
import { pingRolesInThread } from "./helpers/discord_notifications.js";
import { loadDiscordConfig } from "./helpers/discord_config.js";
import { fetchTextWithRetry } from "./helpers/utils.js";
import { formatDate, getIOSInfo as _getIOSInfo, getAndroidInfo } from "./helpers/app_info.js";

// ===== CẤU HÌNH =====
const STATE_FILE = "last_versions.json";
const APPS_SHEET_URL = process.env.APPS_SHEET_URL; // CSV: platform,id,name_fallback

// =====================

function loadState() {
  try {
    if (fs.existsSync(STATE_FILE)) {
      const raw = fs.readFileSync(STATE_FILE, "utf8").trim();
      if (!raw) return {};
      const parsed = JSON.parse(raw);
      for (const k of Object.keys(parsed)) {
        if (typeof parsed[k] === "string") parsed[k] = { version: parsed[k] };
      }
      return parsed;
    }
  } catch (e) { console.log("❌ Lỗi load state:", e.message); }
  return {};
}

function saveState(state) {
  try {
    fs.writeFileSync(STATE_FILE, JSON.stringify(state, null, 2));
    console.log(`💾 Đã lưu state vào ${STATE_FILE}`);
  } catch (e) { console.log("❌ Lỗi lưu state:", e.message); }
}

async function ensureAppThread(taskName, platform, appId, appDisplayName, state) {
  let created = false;
  const entry = state[appId] || {};
  const channelId = pickChannelId(taskName, platform);
  const threadName = `${appDisplayName}`;

  if (entry.thread_id) {
    const check = await ensureThreadBelongsToChannel(entry.thread_id, channelId);
    if (check.ok) {
      const meta = check.info.thread_metadata || {};
      if (meta.locked) {
        console.log("🔒 Thread cũ locked → tạo mới.");
      } else {
        if (meta.archived) await unarchiveThread(entry.thread_id, 10080);
        return { threadId: entry.thread_id, created }; // ✅ reuse
      }
    } else {
      console.log(`↪️ Thread cũ không hợp lệ (${check.reason}) → tạo mới trong channel đúng.`);
    }
  }
  
  const reused = await reuseThreadByNameInThisChannel(channelId, threadName);
  if (reused) {
    state[appId] = { ...(state[appId] || {}), thread_id: reused };
    console.log(`♻️ Dùng lại thread theo tên trong channel: ${threadName} (${reused})`);
    return { threadId: reused, created };
  }
  
  const threadId = await createThreadInTextChannel(channelId, threadName, 10080);
  //await pingRolesInThread(threadId);
  created = true;
  state[appId] = { ...(state[appId] || {}), thread_id: threadId };
  console.log(`🧵 Tạo thread game: ${threadName} (${threadId})`);
  return { threadId, created };
}

async function loadAppsFromSheet(url) {
  try {
    const csvText = await fetchTextWithRetry(url);
    const parsed = Papa.parse(csvText, { header: true, skipEmptyLines: false });
    const cfg = { ios: [], android: [] };
    for (const row of parsed.data) {
      const platform = (row.platform || "").trim().toLowerCase();
      const id = (row.id || "").trim();
      const name_fallback = (row.name_fallback || "").trim() || id;
      if (!platform || !id) continue;
      if (platform === "ios" || platform === "android") {
        cfg[platform].push({ id, name_fallback });
      }
    }
    if (!cfg.ios.length && !cfg.android.length) return null;
    console.log(`✅ Đã tải cấu hình từ Google Sheet: ${cfg.ios.length} iOS, ${cfg.android.length} Android`);
    return cfg;
  } catch (e) {
    console.log("❌ Lỗi đọc Google Sheet CSV:", e.message);
    return null;
  }
}

async function loadAppsConfig() {
  if (APPS_SHEET_URL) {
    const cfg = await loadAppsFromSheet(APPS_SHEET_URL);
    if (cfg) return cfg;
    console.log("⚠️ Sheet lỗi/trống. Dùng APPS_CONFIG nếu có.");
  }
  console.log("⚠️ Dùng cấu hình mặc định cứng.");
  return {
    ios: [{ id: "1517783697", name_fallback: "Genshin Impact" }],
    android: [{ id: "com.miHoYo.GenshinImpact", name_fallback: "Genshin Impact" }]
  };
}

// --- Scrape App Store web ---
function pickFromSrcset(srcset) {
  if (!srcset) return null;

  const parts = srcset.split(",").map((p) => p.trim());
  let best = null;

  for (const part of parts) {
    const [url, size] = part.split(/\s+/);
    const m = size && size.match(/(\d+)w/);
    const w = m ? parseInt(m[1], 10) : 0;

    if (!best || w > best.w) {
      best = { url, w };
    }
  }

  return best ? best.url : null;
}

function normalizeUrl(u, base) {
  if (!u) return null;
  if (u.startsWith("http://") || u.startsWith("https://")) return u;
  try {
    return new URL(u, base).href;
  } catch {
    return null;
  }
}

async function scrapeAppStoreScreenshots(appId, limit = 1) {
  const pageUrl = `https://apps.apple.com/app/id${appId}`;

  const res = await axios.get(pageUrl, {
    timeout: 20000,
    maxRedirects: 5,
    headers: {
      "User-Agent":
        "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/124 Safari/537.36",
      "Accept-Language": "en-US,en;q=0.9",
      "Accept":
        "text/html,application/xhtml+xml,application/xml;q=0.9,*/*;q=0.8",
    },
  });

  const html = res.data;
  const $ = cheerio.load(html);

  const candidates = [];

  const selectors = [
    ".we-screenshot-viewer__screenshots img",
    ".we-screenshot-viewer__screenshots source[srcset]",
    "figure.we-artwork img",
    "figure.we-artwork source[srcset]",
    "picture source[srcset]",
    "img[srcset]",
    "img[src]",
  ];

  $(selectors.join(", ")).each((_, el) => {
    const tag = el.name?.toLowerCase?.() || "";
    const $el = $(el);

    let rawUrl = null;
    if (tag === "source") {
      rawUrl =
        pickFromSrcset($el.attr("srcset") || $el.attr("data-srcset")) || null;
    } else {
      rawUrl =
        $el.attr("src") ||
        $el.attr("data-src") ||
        pickFromSrcset($el.attr("srcset") || $el.attr("data-srcset")) ||
        null;
    }

    const url = normalizeUrl(rawUrl, pageUrl);
    if (!url) return;

    const alt = $el.attr("alt") || "";

    candidates.push({ url, alt });
  });

  const iconPattern =
    /(icon|icons|badge|logo|apple-touch-icon|square|thumbnail|small|100x100|60x60|512x512|1024x1024)/i;
  const screenshotKeywords =
    /(screenshot|screen|iphone|ipad|android|screenshotUrls)/i;

  const scored = candidates.map(({ url, alt }) => {
    let score = 0;

    // + điểm nếu có từ khóa screenshot
    if (screenshotKeywords.test(alt)) score += 50;
    if (screenshotKeywords.test(url)) score += 40;

    // + điểm nếu là ảnh to (dựa trên pattern WxH trong URL)
    const dimMatch = url.match(/(\d{2,4})x(\d{2,4})/);
    if (dimMatch) {
      const w = parseInt(dimMatch[1], 10);
      const h = parseInt(dimMatch[2], 10);
      if (w >= 300 || h >= 300) score += 30;
      if (w >= 600 || h >= 600) score += 10; // ưu tiên siêu to
    }

    // - điểm nếu giống icon / logo nhỏ
    if (iconPattern.test(url) || iconPattern.test(alt)) score -= 60;

    return { url, score };
  });

  // sort theo score giảm dần
  scored.sort((a, b) => b.score - a.score);

  // unique URL + limit
  const out = [];
  for (const s of scored) {
    if (!out.includes(s.url)) {
      out.push(s.url);
      if (out.length >= limit) break;
    }
  }

  return out;
}

// Wrapper: getIOSInfo with screenshot scraping
async function getIOSInfo(appId) {
  return _getIOSInfo(appId, { scrapeScreenshot: scrapeAppStoreScreenshots });
}

async function sendDiscordEmbed(threadId, appName, platform, oldVersion, info) {
  try {
    const embed = {
      title: `📢 ${appName} (${platform}) vừa cập nhật!`,
      url: info.url,
      description:
        `**Platform:** ${platform}\n` +
        `**Old Version:** \`${oldVersion || "N/A"}\`\n` +
        `**New Version:** \`${info.version}\`\n` +
        `**Release Date:** ${info.releaseDate || "Không rõ"}\n` +
        `**Developer:** [${info.developer || "Không rõ"}](${info.developerUrl || info.url})\n` +
        `**Genres:** ${info.genres || "Không rõ"}\n\n` +
        `**Release Notes:**\n${(info.releaseNotes || "").slice(0, 1000)}`,
      color: 0x1abc9c,
      thumbnail: { url: info.icon },
      footer: { text: "App Update Notifier" }
    };
    if (info.screenshot) embed.image = { url: info.screenshot };

    await sendMessageToThread(threadId, { embeds: [embed] });
    console.log(`✅ Đã gửi thông báo Discord cho ${appName} (${platform})`);
  } catch (e) {
    console.log("❌ Lỗi gửi Discord:", e.message);
  }
}

async function main() {
  console.log("🔄 Bắt đầu kiểm tra cập nhật phiên bản ứng dụng...");
  await loadDiscordConfig(); // load channel + role config
  const apps = await loadAppsConfig();
  console.log(`📋 Danh sách ứng dụng: ${apps.ios.length} iOS, ${apps.android.length} Android`);
  const state = loadState();
  const getOldVersion = (entry) => {
    if (!entry) return null;
    if (typeof entry === "string") return entry;
    if (typeof entry.version === "string") return entry.version;
    return null; // có thể đã có thread_id nhưng chưa có version
  };

  let changed = false;

  console.log("📱 Kiểm tra iOS...");
  for (const a of apps.ios) {
    console.log(`   🔍 ${a.name_fallback}...`);
    const info = await getIOSInfo(a.id);
    if (!info) { console.log("    ⚠️ Không lấy được thông tin."); continue; }
    const { threadId: iosThreadId, created: iosCreated } = await ensureAppThread("check_app_updates", "ios", a.id, info.name || a.name_fallback, state);
    const old = getOldVersion(state[a.id]);
    const first = !state[a.id] || !old; // 👈 app này chưa từng có trong state
    const isDifferent = info.version !== old;
    console.log(`    - Cũ: ${old || "N/A"} | Mới: ${info.version} | Khác nhau: ${isDifferent}`);
    if (isDifferent) {
      if (!first) {
        if (!iosCreated){
          await pingRolesInThread(iosThreadId);
        }
        await sendDiscordEmbed(iosThreadId, info.name || a.name_fallback, "iOS", old, info);
      }
      state[a.id] = { ...(state[a.id] || {}), version: info.version };
      changed = true;
    }
  }

  console.log("🤖 Kiểm tra Android...");
  for (const a of apps.android) {
    console.log(`   🔍 ${a.name_fallback}...`);
    const info = await getAndroidInfo(a.id);
    if (!info) { console.log("    ⚠️ Không lấy được thông tin."); continue; }
    const { threadId: androidThreadId, created: androidCreated } = await ensureAppThread("check_app_updates", "android", a.id, info.name || a.name_fallback, state);
    const old = getOldVersion(state[a.id]);
    const first = !state[a.id] || !old;
    const isDifferent = info.version !== old;
    console.log(`    - Cũ: ${old || "N/A"} | Mới: ${info.version} | Khác nhau: ${isDifferent}`);
    if (isDifferent) {
      if (!first) {
        if (!androidCreated){
          await pingRolesInThread(androidThreadId);
        }
         await sendDiscordEmbed(androidThreadId, info.name || a.name_fallback, "Android", old, info);
      }
      state[a.id] = { ...(state[a.id] || {}), version: info.version };
      changed = true;
    }
  }

  saveState(state);
  if (!changed) {
    console.log("ℹ️ Không phát hiện cập nhật phiên bản nào");
  }
}

main().catch(e => { console.error(e); process.exit(1); });
